mock-hpos-api/index.js

```diff
@@ -0,0 +1,147 @@
+const express = require('express')
+const cors = require('cors')
+const bodyParser = require('body-parser')
+const stringify = require('fast-json-stable-stringify')
+const _ = require('lodash')
+const HpAdminKeypair = require('@holo-host/hp-admin-keypair').HpAdminKeypair
+const { signPayload, hashString } = require('./authUtils')
+const defaultResponse = require('./defaultResponse')
+
+const HC_PUBKEY = '5m5srup6m3b2iilrsqmxu6ydp8p8cr0rdbh4wamupk3s4sxqr5'
+
+const API_ROOT = '/api/v1'
+
+function removeApiRoot (path) {
+  if (path.indexOf(API_ROOT) == 0) {
+    return path.slice(API_ROOT.length);
+  } else {
+    return path
+  }
+}
+
+function generateResponseKey (method, path, data) {
+  return `${method}:${removeApiRoot(path)}:${JSON.stringify(data)}`
+}
+
+// NEXT_PATH is used to key a next response in place of an actual path. It's required to be different from any value that could come from req.path
+const NEXT_PATH = ' . next . '
+
+const NEXT_RESPONSE_KEY = generateResponseKey('', NEXT_PATH, {})
+
+class MockHposApi {
+  constructor(port, authEmail, authPassword) {
+    this.port = port
+    this.shouldCheckAuth = !!authEmail && !!authPassword
+    this.authEmail = authEmail
+    this.authPassword = authPassword
+    
+    this.initializeResponses()
+  }
+
+  static async start (port, authEmail, authPassword) {
+    const mockHposApi = new MockHposApi(port, authEmail, authPassword)
+
+    const app = express()
+
+    app.use(cors())
+    app.use(bodyParser.json())
+    app.use(mockHposApi.checkAuth.bind(mockHposApi))
+    app.use(mockHposApi.handleRequest.bind(mockHposApi))
+
+    await mockHposApi.startServer(app)
+
+    return mockHposApi
+  }
+
+  // this is more convoluted than I'd like because I want init to be a blocking call, so have to uncallback app.listen
+  startServer (app) {
+    return new Promise((resolve, reject) => {
+      this.server = app.listen(this.port, () => {
+        console.log(`Mock hpos-admin-api listening at http://localhost:${this.port}`)
+        resolve()
+      }).on('error', reject)
+    })
+  }
+
+  async checkAuth (req, res, next) {
+    if (!this.shouldCheckAuth) {
+      next()
+    } else {
+
+      const { method, path, body } = req
+
+      const keypairInstance = new HpAdminKeypair(HC_PUBKEY, this.authEmail, this.authPassword)
+
+      const signature = await signPayload(keypairInstance, method, path, body)
+
+      if (signature === req.header('x-hpos-admin-signature')) {
```

Comment:
Is this the right way to test signatures? It feels vaguely tautological that we implement the same key signing in both the test and the source and then check that they match.

Should we instead be trying to implement the verification that the crypto-server do? Or perhaps that would defeat the point of a mock and be moving into integration testing?

Reply:
unit tests are often tautological. It's just more obvious here because of the hoops we have to jump through. But saying that, it makes sense to simulate the server as much as possible here. Just had a look in `holo-nixpkgs` and I can't find that logic. Is it in there? Can you point me in the right direction?

Reply:
https://github.com/Holo-Host/hp-admin-crypto/blob/4252d1c3879719d119869d5792b3256da04029f7/server/src/main.rs#L111-L124

To be clear - I am not requesting a change here as I don't know enough about test design. Was just a comment

Reply:
👍 

Thought about this more. What we would be doing in that case is testing the `@holo-host/hp-admin-keypair` library, not the `host-console-ui` code. Presumably that lib has its own tests, and if not we should add them there.